README.md
## INITIALIZE

- install nodejs
- install [gt](https://github.com/vivaxy/granturismo)

    `yarn global add granturismo`

- add scaffold into gt `gt config add vivaxy/gt-react-scaffold https://github.com/vivaxy/gt-react-scaffold.git`
- create your project directory `mkdir my-project-name && cd my-project-name` and `git clone ...`
- run `gt init`
- select `vivaxy/gt-react-scaffold`

## CONTRIBUTE

Feel free to submit any issue.

Steps to make contributions

- update codes
- submit your changes by [`gacp`](https://github.com/vivaxy/gacp)
- run `yarn run release` to update `CHANGELOG.md` and bump version
- run [`gacp`](https://github.com/vivaxy/gacp) to push changes into origin

----------

# gt-react-scaffold

## FEATURE

- react
- redux
- webpack
- eslint
- page navigation animation
- react router
- react redux
- redux thunk
- react hot reload
- webpack development server
- multiple entries
- mock server
- unified error center
- internationalization
- environments
- webpack visualizer
- postcss
- autoprefixer
- less
- test cases

## STRUCTURE

![flowchart](./docs/flowchart.png)

As we find out html are mostly same in react projects, we extract the same html into a template file which lays in `html` folder.

`html-webpack-plugin` is used to generate released html files, which load corresponding js files.

`entries` stores js main methods, we split main container to `containers` because entry could not be hot-module-replaced.

`entries`s import `containers`s, which is one to one correspondent. `render` method in entry provides all commonly used setups, such as redux store, provider, injectTapEventPlugin for material-ui, and custom styles.

`containers` is the headquarters which imports all dummy components, passing through props to them. Also, `containers` interacts with redux `reducers` and `actions`.

It is recommended that you wrap containers into another. `redux` will re-render a container when and only when mapped reducers have been updated.

We do not use async redux for ajax requests because ajax status is not that important for us to record. So, we invoke `api` in `containers`, get response, and determine what `actions` to call.

We store every string in `i18n` for better internationalization.

Built project files will be in the `release` folder. With `build.log`, you can see every build details. With `stats.html`, you can optimize your dependencies to make project smaller.

## DEVELOP

- `yarn run dev`

## BUILD

- `yarn run build`

## REFERENCE

- https://github.com/webpack/docs/wiki/webpack-dev-server
- https://github.com/gaearon/react-hot-loader/tree/master/docs
- https://github.com/reactjs/react-router-redux
- https://github.com/ampedandwired/html-webpack-plugin
- https://github.com/ReactTraining/react-router


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.8.0"></a>
# [1.8.0](https://github.com/vivaxy/gt-react-scaffold/compare/v1.7.0...v1.8.0) (2017-05-04)


### Features

* **dependencies:** :arrow_up:Move react react-dom redux redux-thunk into dependencies ([20b4304](https://github.com/vivaxy/gt-react-scaffold/commit/20b4304))
* **report:** :sparkles:Replace `webpack-visualizer-plugin` with `webpack-bundle-analyzer` ([8057b69](https://github.com/vivaxy/gt-react-scaffold/commit/8057b69))



<a name="1.7.0"></a>
# [1.7.0](https://github.com/vivaxy/gt-react-scaffold/compare/v1.6.1...v1.7.0) (2017-05-03)


### Features

* **babel:** :sparkles:Using babel-runtime instead of babel-polyfill, reduce the release files' size ([f2b729a](https://github.com/vivaxy/gt-react-scaffold/commit/f2b729a))



<a name="1.6.1"></a>
## [1.6.1](https://github.com/vivaxy/gt-react-scaffold/compare/v1.6.0...v1.6.1) (2017-05-03)



<a name="1.6.0"></a>
# [1.6.0](https://github.com/vivaxy/gt-react-scaffold/compare/v1.5.2...v1.6.0) (2017-04-18)


### Features

* **dep:** :alien:Upgrade react to 15.5, and react-dom to 15.5 ([230e3e1](https://github.com/vivaxy/gt-react-scaffold/commit/230e3e1))
* **eslint:** :sparkles:Add eslint-loader, Upgrade react-pianist and react-router because of prop-ty ([7d695c3](https://github.com/vivaxy/gt-react-scaffold/commit/7d695c3))
* **openBrowser:** :sparkles:Optimize open browser ([cb55563](https://github.com/vivaxy/gt-react-scaffold/commit/cb55563))



<a name="1.5.2"></a>
## [1.5.2](https://github.com/vivaxy/gt-react-scaffold/compare/v1.5.1...v1.5.2) (2017-04-10)



<a name="1.5.1"></a>
## [1.5.1](https://github.com/vivaxy/gt-react-scaffold/compare/v1.5.0...v1.5.1) (2017-04-01)


### Bug Fixes

* **gt:** :bug:Fix gt after message; Add yarn install in gt; Update webpack-dev-server to 2.x in pack ([8849562](https://github.com/vivaxy/gt-react-scaffold/commit/8849562))
* **gt:** :bug:Fix gt.js ([e84665d](https://github.com/vivaxy/gt-react-scaffold/commit/e84665d))



<a name="1.5.0"></a>
# [1.5.0](https://github.com/vivaxy/gt-react-scaffold/compare/v1.4.0...v1.5.0) (2017-03-04)


### Features

* **test:** :sparkles:Add ava test ([a5bf82f](https://github.com/vivaxy/gt-react-scaffold/commit/a5bf82f))



<a name="1.4.0"></a>
# [1.4.0](https://github.com/vivaxy/gt-react-scaffold/compare/v1.3.0...v1.4.0) (2017-03-04)


### Bug Fixes

* :bug:Fix yarn.lock resolved urls; Add missing dependency: redux; ([229ac48](https://github.com/vivaxy/gt-react-scaffold/commit/229ac48))
* **gt:** :bug:Fix copy `.npmrc` ([baf9a1a](https://github.com/vivaxy/gt-react-scaffold/commit/baf9a1a))
* **gt:** :bug:Fix template, render file project name ([9b72633](https://github.com/vivaxy/gt-react-scaffold/commit/9b72633))
* **gt.js:** :bug:Fix gt copy file ([5c24c19](https://github.com/vivaxy/gt-react-scaffold/commit/5c24c19))
* **gt.js:** :bug:Update gt.js ([565f05d](https://github.com/vivaxy/gt-react-scaffold/commit/565f05d))
* **scripts:** :bug:Fix release script ([04ece5d](https://github.com/vivaxy/gt-react-scaffold/commit/04ece5d))


### Features

* **babel:** :sparkles:Update babel-preset-es2015 babel-preset-latest ([5545282](https://github.com/vivaxy/gt-react-scaffold/commit/5545282))
* **eslint:** :sparkles:Add eslint, rules based on airbnb/javascript ([2b976e2](https://github.com/vivaxy/gt-react-scaffold/commit/2b976e2))
* **npmrc:** :art:Add .npmrc to maintain npm registry ([5dac42b](https://github.com/vivaxy/gt-react-scaffold/commit/5dac42b))



<a name="1.3.0"></a>
# [1.3.0](https://github.com/vivaxy/gt-react-scaffold/compare/v1.2.0...v1.3.0) (2017-01-26)


### Bug Fixes

* **gt.js:** :bug:Fix gt info in README.md ([00d772b](https://github.com/vivaxy/gt-react-scaffold/commit/00d772b))


### Features

* **postcss:** :sparkles:Add postcss ([df5d83a](https://github.com/vivaxy/gt-react-scaffold/commit/df5d83a))
* **rename:** :sparkles:Rename `react-scaffold` to `gt-react-scaffold` ([5107d7b](https://github.com/vivaxy/gt-react-scaffold/commit/5107d7b))
* **webpack:** :sparkles:Use `scripts/dev.js` to start webpack dev server, for more controllable. No ([f6b7103](https://github.com/vivaxy/gt-react-scaffold/commit/f6b7103))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/vivaxy/gt-react-scaffold/compare/v1.1.0...v1.2.0) (2017-01-09)


### Bug Fixes

* **gt:** :bug:fix source folder name ([23590ca](https://github.com/vivaxy/gt-react-scaffold/commit/23590ca))
* **yarn:** :arrow_up: ([e314496](https://github.com/vivaxy/gt-react-scaffold/commit/e314496))


### Features

* **fetch:** :art:remove fetch cache ([490b408](https://github.com/vivaxy/gt-react-scaffold/commit/490b408))
* **fetch:** :sparkles:add timeout for fetch ([8a0cddb](https://github.com/vivaxy/gt-react-scaffold/commit/8a0cddb))
* :sparkles:improve page transition style ([3552fc6](https://github.com/vivaxy/gt-react-scaffold/commit/3552fc6))
* **gt:** :sparkles:add sleep in gt.js ([e15b49c](https://github.com/vivaxy/gt-react-scaffold/commit/e15b49c))
* **page transition:** :sparkles: ([0643b9e](https://github.com/vivaxy/gt-react-scaffold/commit/0643b9e))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/vivaxy/gt-react-scaffold/compare/v1.0.2...v1.1.0) (2016-11-27)


### Bug Fixes

* **fetch:** :bug:fix cached response cannot  more than once ([7883557](https://github.com/vivaxy/gt-react-scaffold/commit/7883557))
* **webpack:** :bug:only open browser when dev server ([9236361](https://github.com/vivaxy/gt-react-scaffold/commit/9236361))


### Features

* **fetch:** :art:support CORS ([1ce3a82](https://github.com/vivaxy/gt-react-scaffold/commit/1ce3a82))
* **webpack:** :sparkles:added OccurrenceOrderPlugin ([e003181](https://github.com/vivaxy/gt-react-scaffold/commit/e003181))

src/lib/store.js
/**
 * @since 2016-09-16 18:11
 * @author vivaxy
 */

import { createStore, combineReducers, compose, applyMiddleware } from 'redux';
import { routerMiddleware } from 'react-router-redux';
import { hashHistory } from 'react-router';
import thunk from 'redux-thunk';

import reducers from '../redux/reducers';

const routing = routerMiddleware(hashHistory);

const store = createStore(combineReducers(reducers), compose(
    applyMiddleware(thunk, routing),
    window.devToolsExtension ? window.devToolsExtension() : (f) => f, // eslint-disable-line arrow-body-style
));

if (module.hot) {
    // Enable Webpack hot module replacement for reducers
    module.hot.accept('../redux/reducers', () => {
        const nextRootReducer = require('../redux/reducers').default; // eslint-disable-line global-require
        store.replaceReducer(combineReducers(nextRootReducer));
    });
}

export default store;
